### PR TITLE
EDGE3D App

### PR DESCRIPTION
**Pull request by pguthrey**
# The EDGE3D kernel computes the summed mass + stiffness matrix of low order edge bases, which is relevant to MHD discretizations.

- This PR is a feature
- It does the following (modify list as needed):
  - Adds the EDGE3D app

### CODE REVIEW

File: src/apps/EDGE3D.cpp
Comment (artv3) [inline src/apps/EDGE3D.cpp:67]: I believe this is used to scale terms in the output array in a manner that helps avoid getting the right answer by accident. I'm forgetting the details, but perhaps @rhornung67  can help explain a bit.

File: src/apps/EDGE3D.hpp
Comment (artv3) [inline src/apps/EDGE3D.hpp:511]: Perhaps delete if not used?

**Review by artv3 — APPROVED** [review]

File: src/apps/EDGE3D.cpp
Comment (rhornung67) [reply src/apps/EDGE3D.cpp:67]: Depending on a kernel's properties (default problem size, default # samples, how a checksum is calculated, typical magnitude of result values used to calculate checksums, etc.) we sometimes ad a scale factor to the checksum calculation to avoid overflow and accuracy issues. Typically, we try to shoot for checksum differences < 1e-8 between variants for correct results.

File: src/apps/EDGE3D-Cuda.cpp
Comment (CRobeck) [inline src/apps/EDGE3D-Cuda.cpp:57]: Why do we need two indexes here? `ii` appears to be unused.

Comment (pguthrey) [reply src/apps/EDGE3D-Cuda.cpp:57]: i = ii + ibegin ?

File: src/apps/EDGE3D.hpp
Comment (pguthrey) [reply src/apps/EDGE3D.hpp:511]: Any idea what this should be? I am inclined to uncomment this default value and delete what I used.

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [reply src/apps/EDGE3D-Cuda.cpp:57]: I think the suggestion here is `Index_type i = ibegin + blockIdx.x * block_size + threadIdx.x;`

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D.cpp:73]: Where are the lambda variants?

File: src/apps/EDGE3D.hpp
Comment (MrBurmark) [inline src/apps/EDGE3D.hpp:9]: where's the loop over zones?

File: src/apps/EDGE3D.hpp
Comment (MrBurmark) [inline src/apps/EDGE3D.hpp:252]: do we need to rename i to iz?

**Comment by pguthrey** [comment]
Todo: investigate launch bounds.

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Cuda.cpp:40]: Hmm, already have launch bounds.

File: src/apps/EDGE3D-Hip.cpp
Comment (artv3) [inline src/apps/EDGE3D-Hip.cpp:40]: launch bounds is on the HIP on too.. weird..

File: src/apps/EDGE3D.hpp
Comment (MrBurmark) [inline src/apps/EDGE3D.hpp:516]:
```suggestion
  using gpu_block_sizes_type = gpu_block_size::make_list_type<default_gpu_block_size>;
```

File: src/apps/EDGE3D.hpp
Comment (MrBurmark) [inline src/apps/EDGE3D.hpp:196]: let's consider whether we think this should be in the macro or in a function

Comment (rhornung67) [reply src/apps/EDGE3D.hpp:196]: Also, it should be changed because it doesn't compile.

Comment (pguthrey) [reply src/apps/EDGE3D.hpp:196]: Ah those pesky non-gpu builds... ;)

fixed

File: src/apps/ahelper.hpp
Comment (pguthrey) [reply src/apps/ahelper.hpp]: changed to "mixed_fem_helper"

File: src/apps/EDGE3D.cpp
Comment (pguthrey) [reply src/apps/EDGE3D.cpp:73]: Fixed

**Comment by rhornung67** [comment]
@pguthrey I fixed all the issues in your PR and got all the checks to pass. I would like @MrBurmark to do a final review before merging.

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D.cpp:52]: Is this just the size of loaded data, or loaded and stored data size?

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D.cpp:74]:
```suggestion
  setVariantDefined( Base_Seq );
  setVariantDefined( Lambda_Seq );
  setVariantDefined( RAJA_Seq );
```

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D.cpp:87]: what about lambda variants here?

File: src/apps/mixed_fem_helper.hpp
Comment (MrBurmark) [inline src/apps/mixed_fem_helper.hpp:9]: Shall we use ifdefs, I don't think we've used pragma once elsewhere in this code.

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [reply src/apps/EDGE3D.cpp:52]: Is this comment accurate?

Comment (pguthrey) [reply src/apps/EDGE3D.cpp:52]: Let's go through the impl and determine that

File: src/apps/EDGE3D.hpp
Comment (MrBurmark) [reply src/apps/EDGE3D.hpp]: Is this resolved?

File: src/apps/EDGE3D.cpp
Comment (pguthrey) [reply src/apps/EDGE3D.cpp]: Okay so now everything has lambda variants. I will monitor testing to see if it passes.

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Cuda.cpp:88]:
```suggestion
        EDGE3D_BODY;
```

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Cuda.cpp:63]:
```suggestion
```

File: src/apps/EDGE3D-Hip.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Hip.cpp:63]:
```suggestion
```

File: src/apps/EDGE3D-Hip.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Hip.cpp:88]:
```suggestion
        EDGE3D_BODY;
```

**Review by MrBurmark — CHANGES_REQUESTED** [review]
Extra lambdas in GPU variants

File: src/apps/EDGE3D.hpp
Comment (CRobeck) [inline src/apps/EDGE3D.hpp:196]: RAJA_INLINE

File: src/apps/EDGE3D.hpp
Comment (CRobeck) [inline src/apps/EDGE3D.hpp:197]: double->Real_Type
int ->Index_Type

File: src/apps/mixed_fem_helper.hpp
Comment (CRobeck) [inline src/apps/mixed_fem_helper.hpp:218]: double->Real_Type

File: src/apps/mixed_fem_helper.hpp
Comment (CRobeck) [inline src/apps/mixed_fem_helper.hpp:261]: Would this make more sense as constexpr instead of a non type template parameter?

File: src/apps/mixed_fem_helper.hpp
Comment (CRobeck) [inline src/apps/mixed_fem_helper.hpp:233]: Would this make more sense as constexpr instead of a non type template parameter?

File: src/apps/mixed_fem_helper.hpp
Comment (CRobeck) [inline src/apps/mixed_fem_helper.hpp:374]: 1. Is the cast to void just for compiler warning? 
2. Can we remove the y and z from this scope block?
3. Is the bottom section of this scope block unreachable?

File: src/apps/EDGE3D.cpp
Comment (pguthrey) [inline src/apps/EDGE3D.cpp:47]:
```suggestion
```

**Comment by rhornung67** [comment]
@MrBurmark @pguthrey I merged develop into this, fixed build issues, and addressed several issues in reviewer comments. Is it good to merge?

**Comment by pguthrey** [comment]
> @MrBurmark @pguthrey I merged develop into this, fixed build issues, and addressed several issues in reviewer comments. Is it good to merge?

My plan is to finish this MR during the hackathon this week. I'll let you know once all the interested parties are happy.

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp:376]: 1.  Yes
3.  The return statement is the last line that is called.

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp]: Thanks for that @rhornung67

**Comment by pguthrey** [comment]
Ah I see the spelling was fixed in
https://github.com/LLNL/RAJAPerf/commit/0987b08e1695b9a637fa8086839ab93b7fda8c96

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Cuda.cpp:66]: Need to add stream/resource use to all of the cuda/hip kernels, see the other cuda/hip kernels for an example.

Comment (pguthrey) [reply src/apps/EDGE3D-Cuda.cpp]: Done

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp]: Done

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp]: Resolved in person

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp]: Resolved in person

File: src/apps/mixed_fem_helper.hpp
Comment (pguthrey) [reply src/apps/mixed_fem_helper.hpp]: Done

File: src/apps/EDGE3D-OMP.cpp
Comment (CRobeck) [inline src/apps/EDGE3D-OMP.cpp:59]: Can we be consistent between RepIndex_type and Index_type in these loops?

Comment (rhornung67) [reply src/apps/EDGE3D-OMP.cpp:59]: Thanks for catching that @CRobeck

**Comment by pguthrey** [comment]
Something wrong with Lambda_HIP. Will investigate.

File: src/apps/EDGE3D-Cuda.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D-Cuda.cpp:89]:
```suggestion
        ibegin, iend, edge3d_lam);
```

File: src/apps/EDGE3D-OMP.cpp
Comment (MrBurmark) [reply src/apps/EDGE3D-OMP.cpp:59]: Did you add an issue for this?

File: src/apps/EDGE3D.cpp
Comment (MrBurmark) [inline src/apps/EDGE3D.cpp:50]: I like the idea of using named variables here, big improvement in readability. I'll add an issue to refactor this for the rest of the perf suite.
```suggestion
  size_t reads_per_node = 3*sizeof(Real_type);
  size_t writes_per_zone = 1*sizeof(Real_type);
  setBytesPerRep( writes_per_zone * getItsPerRep() +
                  reads_per_node * (getItsPerRep() + 1+m_domain->jp+m_domain->kp) );
```

**Comment by pguthrey** [comment]
@MrBurmark I think this is good to go.

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
We can merge this when checks come back all green.

**Comment by rhornung67** [comment]
@MrBurmark @pguthrey checks are all green. Merge this?

**Pull request merged**
